src/dataframe/dataframe.jl

```diff
@@ -396,25 +415,33 @@ function Base.setindex!(df::DataFrame, v, col_ind::ColumnIndex)
 end
 
 # df[MultiColumnIndex] = DataFrame
-function Base.setindex!(df::DataFrame, new_df::DataFrame, col_inds::AbstractVector{Bool})
+function Base.setindex!(df::DataFrame, new_df::DataFrame,
+                        col_inds::AbstractVector{<:Union{Bool,Missing}})
+    length(col_inds) != ncol(df) && throw(BoundsError(df, col_inds))
     setindex!(df, new_df, findall(col_inds))
 end
 function Base.setindex!(df::DataFrame,
                         new_df::DataFrame,
                         col_inds::AbstractVector{<:ColumnIndex})
+    # it is allowed that new_df has more columns than required
     for j in 1:length(col_inds)
         insert_single_column!(df, new_df[j], col_inds[j])
     end
     return df
 end
 
 # df[MultiColumnIndex] = AbstractVector (REPEATED FOR EACH COLUMN)
-function Base.setindex!(df::DataFrame, v::AbstractVector, col_inds::AbstractVector{Bool})
+function Base.setindex!(df::DataFrame, v::AbstractVector,
+                        col_inds::AbstractVector{<:Union{Bool,Missing}})
+    length(col_inds) != ncol(df) && throw(BoundsError(df, col_inds))
     setindex!(df, v, findall(col_inds))
 end
 function Base.setindex!(df::DataFrame,
                         v::AbstractVector,
                         col_inds::AbstractVector{<:ColumnIndex})
+    # this allows to add arbitrarily many columns to a DataFrame, e.g.
```

Comment:
That doesn't sound too bad. I'm not sure there are similar structures in Julia where it would make sense to use the same approach (dicts can't because ranges are valid keys).

Reply:
OK - leaving it, but it should be documented.